File: leetcode/longest_consecutive_sequence.py
# Given an unsorted array of integers, find the length of the longest 
# consecutive elements sequence.
# Time Complexity: O(n) in average
# Space Complexity: O(n)

class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def longestConsecutive(self, nums):
        b, result = {}, 0
        for num in nums:
            if num not in b:
                left, right = b.get(num - 1, num), b.get(num + 1, num)
                result = max(result, right - left + 1)
                b[num], b[left], b[right] = num, right, left
        return result

if __name__ == "__main__":
    solution = Solution()
    print solution.longestConsecutive([100, 4, 200, 1, 3, 2])

File: leetcode/longest_consecutive_sequence.js
/* Given an unsorted array of integers, find the length of the longest 
 * consecutive elements sequence.
 * Time Complexity: O(n) in average
 * Space Complexity: O(n)
 */

/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function(nums) {
    var b = new Map(), result = 0;
    for (var num of nums) {
        if (!b.has(num)) {
            var left = b.has(num - 1) ? b.get(num - 1) : num;
            var right = b.has(num + 1) ? b.get(num + 1) : num;
            result = Math.max(result, right - left + 1);
            b.set(num, num);
            b.set(left, right);
            b.set(right, left);
        }
    }
    return result;
};
